Matchismo/Model/CardMatchingGame.m

```diff
@@ -63,30 +73,91 @@ - (void)chooseCardAtIndex:(NSUInteger)index
 {
     Card *card = [self cardAtIndex:index];
 
-    if (!card.isMatched) {
-        if (card.isChosen) {
-            card.chosen = NO;
+    if (card.isMatched) {
+        NSLog(@"%@ card is already matched. Ignoring", card.contents);
+        return;
+    }
+
+    if (card.isChosen) {
+        NSLog(@"%@ card is flipped back", card.contents);
+        card.chosen = NO;
+        self.state = @"";
+        return;
+    }
+
+    NSLog(@"%@ card is chosen", card.contents);
+
+    NSArray<Card *> *chosenCards = [self getChosenCards];
+    NSUInteger currentlySelected = chosenCards.count + 1;
+
+    if (currentlySelected == [self requiredToBeChosen]) {
+        NSLog(@"Matching…");
+        NSInteger matchScore = [card match:chosenCards];
+
+        if (matchScore > 0) {
+            NSInteger receivedScore = matchScore * self.matchBonus;
+            self.score += receivedScore;
+            NSArray *matchedCards = [chosenCards arrayByAddingObject:card];
+            [self markAsMatchedCards:matchedCards];
+            self.state = [NSString stringWithFormat:@"Matched %@ for %ld points", [CardMatchingGame makeDescriptionForCards:matchedCards], receivedScore];
         }
         else {
-            for (Card *otherCard in self.cards) {
-                if (otherCard.isChosen && !otherCard.isMatched) {
-                    NSInteger matchScore = [card match:@[otherCard]];
-                    if (matchScore > 0) {
-                        self.score += matchScore * MATCH_BONUS;
-                        card.matched = YES;
-                        otherCard.matched = YES;
-                    }
-                    else {
-                        self.score -= MISMATCH_PENALTY;
-                        otherCard.chosen = NO;
-                    }
-                    break;
-                }
-            }
-            self.score -= COST_TO_CHOOSE;
-            card.chosen = YES;
+            NSArray *mismatchedCards = [chosenCards arrayByAddingObject:card];
+            NSInteger receivedPenalty = self.mismatchPenalty;
+            self.score -= receivedPenalty;
+            [self deselectCards:chosenCards];
+            self.state = [NSString stringWithFormat:@"%@ don't match! %ld points penalty!", [CardMatchingGame makeDescriptionForCards:mismatchedCards], receivedPenalty];
+        }
+    }
+    else {
+        NSArray *selectedCards = [chosenCards arrayByAddingObject:card];
```

Comment:
it is always gonna be just one card that was last selected, isn't it?

Reply:
I assume yes, why?

Reply:
so you don't need another array, you can just use card